src/main/java/emu/grasscutter/game/inventory/Inventory.java

```diff
@@ -138,29 +135,30 @@ public void addItems(Collection<GameItem> items) {
         this.addItems(items, null);
     }
 
-    public void addItems(Collection<GameItem> items, ActionReason reason) {
-        List<GameItem> changedItems = new LinkedList<>();
-
-        for (GameItem item : items) {
-            GameItem result = putItem(item);
-
-            if (result != null) {
-                getPlayer().getBattlePassManager().triggerMission(WatcherTriggerType.TRIGGER_OBTAIN_MATERIAL_NUM, result.getItemId(), result.getCount());
-                changedItems.add(result);
+    public synchronized void addItems(Iterable<GameItem> costItems, int quantity, ActionReason reason) {
```

Comment:
Why `costItems`? They're being received by the inventory, not paid from it. Also I don't see the point of adding a `quantity` field if all existing usage of it just uses the individual counts.

Reply:
I wanted to use `quantity` field at first,however I changed my code later so `quantity` field wasn't used.However,I think it migtht be useful in the future,so I didn't remove it.`costItems` is my fault and I have updated it.

Reply:
I feel like there's just too many options for quantity at that point. The caller is constructing a list of `GameItem`s including individual quantities, there's no need for this function to be unpacking the `GameItem`s and creating new ones just in case something wants to not pre-multiply the quantity (which I'd consider bad form for the caller).